### PR TITLE
Ask to join private channel [IN PROGRESS]

### LINKED ISSUE #295: Anfrage an einen Kanal senden
Sehr geehrte Damen und Herren,

ich bin Assistify user und als User wünsche ich mir, dass ich eine Möglichkeit hätte eine Anfrage auf Freischaltung zu einem Canal schicken zu können. Zur Zeit, kann ich mich irren, gibt es solche Funktionalität nicht. Ich muss erst den/die Canal Owner/in per Tel., Email, persönlich usw. nach fragen, ob er/sie mich freischalten kann. Manchmal sind diese Owners vergeßlich und ich muss die Anfrage oft wiederholen bis ich freigeschaltet werde ;-)(. Wie sehen Sie das?

Mit freundlichen Grüßen
Zamy Getachew

### PR DESCRIPTION
**Pull request by ThomasRoehl**
this closes #295(Anfrage an einem channel senden) and #336(Hide room from Directory)

### CODE REVIEW

**Comment by ThomasRoehl** [comment]
Current Bugs:
- [ ] when clicking on the accept button, the javascript method is not triggered sometimes
- [ ] when clicking on the accept button, the javascript method is triggered twice
- [ ] notification message to the user that has been accepted doesn't find the user mention

File: packages/rocketchat-message-attachments/client/fieldTypeJoinRoomRequest.js [hunk 1]
```diff
@@ -1,9 +1,38 @@
 import {registerFieldTemplate} from './renderField';
 
-registerFieldTemplate('joinRoomRequest', 'JoinRoomRequest');
 
 Template.JoinRoomRequest.helpers({
 	takeAction(status) {
 		return status === 'pending';
+	},
+	hasPermission() {
+		return RocketChat.authz.hasAtLeastOnePermission('add-user-to-any-p-room');
```
Comment (mrsimpson) [inline packages/rocketchat-message-attachments/client/fieldTypeJoinRoomRequest.js]: or `add-user-to-joined-room`

File: packages/rocketchat-message-attachments/client/renderField.js [hunk 2]
```diff
@@ -1,7 +1,37 @@
...
+	// While doing that, add the fieldType as class selector to the events function in order to avoid naming clashes
+	if (events != null) {
+		const uniqueEvents = {};
+		//rename the event handlers so they are unique in the "parent" template to which the events bubble
+		for (const property in events) {
+			if (events.hasOwnProperty(property)) {
+				const event = property.substr(0, property.indexOf(' '));
+				const selector = property.substr(property.indexOf(' ') + 1);
+				Object.defineProperty(uniqueEvents,
+					`${ event }` + ` .${ fieldType } ` + `${ selector }`,
```
Comment (mrsimpson) [inline packages/rocketchat-message-attachments/client/renderField.js]: You don't need the concatenation of multiple templates: `${ event } .${ fieldType } ${ selector }` should do

File: packages/rocketchat-ui/client/views/app/directory.js [hunk 3]
```diff
@@ -32,7 +32,7 @@ function directorySearch(config, cb) {
 				};
 			}
 
-			if (config.type === 'private') {
+			if (config.type === 'private' && RocketChat.authz.hasAllPermission('view-p-room')) {
```
Comment (mrsimpson) [inline packages/rocketchat-ui/client/views/app/directory.js]: I believe it better be `view-outside-room` - either replacing `view-p-room` or with an `&&`

File: packages/rocketchat-i18n/i18n/en.i18n.json [hunk 4]
```diff
@@ -1658,6 +1658,8 @@
   "Report_this_message_question_mark": "Report this message?",
   "Request": "Send Request",
   "Request_to_join_room": "You are not a member of this #<strong>__roomName__</strong> room. Would you like to send request to join",
+  "Request_to_join_room_comment": "A reason, you would like to join this room",
```
Comment (mrsimpson) [inline packages/rocketchat-i18n/i18n/en.i18n.json]: Proposal: The reason why would you like to join this room

de: Der Grund, warum Du dem Kanal beitreten möchtest

File: packages/rocketchat-i18n/i18n/en.i18n.json [hunk 5]
```diff
@@ -1658,6 +1658,8 @@
   "Report_this_message_question_mark": "Report this message?",
   "Request": "Send Request",
   "Request_to_join_room": "You are not a member of this #<strong>__roomName__</strong> room. Would you like to send request to join",
```
Comment (mrsimpson) [inline packages/rocketchat-i18n/i18n/en.i18n.json]: More grammar, please
You are not a member of this #<strong>__roomName__</strong> room.  Would you like to send *a* request for joining it?

de: Du bist bisher nicht Mitglied des Kanals. Möchtest Du fragen, ob Du ihm beitreten darfst?

File: packages/rocketchat-i18n/i18n/en.i18n.json [hunk 6]
```diff
@@ -1658,6 +1658,8 @@
...
+  "Request_to_join_room_comment": "A reason, you would like to join this room",
+  "Request_to_join_room_confirm": "This request to join this room is subject to approval. You can add a reason to support this join request",
```
Comment (mrsimpson) [inline packages/rocketchat-i18n/i18n/en.i18n.json]: This request to join this room is subject to approval. A reason helps the owner of the room to understand why you want to join.

de: Der Eigentümer des Kanals oder ein Moderator muss Deiner Anfrage zustimmen. Wenn Du einen Grund angibst, kann er leichter verstehen, warum Du beitreten möchtest.

File: packages/rocketchat-lib/server/methods/joinRoom.js [hunk 7]
```diff
@@ -74,17 +74,19 @@ Meteor.methods({
 		message.attachments[0].fields[0].responder = responder;
 		return RocketChat.models.Messages.setMessageAttachments(message._id, message.attachments);
 	},
-	getJoinRoomStatus(roomId, user) {
-
+	getJoinRoomStatus(roomName) {
+		check(roomName, String);
+		const room = RocketChat.models.Rooms.findOneByName(roomName);
+		return RocketChat.models.Messages.findLatestJoinRequestByRoom('join-room-request', room._id, Meteor.user().username).fetch();
```
Comment (mrsimpson) [inline packages/rocketchat-lib/server/methods/joinRoom.js]: `join-room-request` should not be a parameter of this method.
Either the find method should be named `findLatestByTypeAndRoomId` or this selector should be hard-coded inside the method.

File: packages/rocketchat-lib/server/models/Messages.js [hunk 8]
```diff
@@ -318,10 +320,34 @@ RocketChat.models.Messages = new class extends RocketChat.models._Base {
 		return this.insert(record);
 	}
 
+	findJoinRequestsByRoom(msgType, rid) {
```
Comment (mrsimpson) [inline packages/rocketchat-lib/server/models/Messages.js]: Also here: Name does not match signature.

File: packages/rocketchat-message-attachments/client/renderField.js (hunk 2 above)
Comment (vickyokrm) [reply packages/rocketchat-message-attachments/client/renderField.js]: changed as mentioned.

File: packages/rocketchat-lib/server/methods/joinRoom.js (hunk 7 above)
Comment (vickyokrm) [reply packages/rocketchat-lib/server/methods/joinRoom.js]: renamed as `findLatestJoinRequestByTypeAndRoomAndUser`

File: packages/rocketchat-lib/server/models/Messages.js (hunk 8 above)
Comment (vickyokrm) [reply packages/rocketchat-lib/server/models/Messages.js]: renamed as `findJoinRequestByMsgTypeAndRoom`

File: packages/rocketchat-i18n/i18n/en.i18n.json (hunk 6 above)
Comment (vickyokrm) [reply packages/rocketchat-i18n/i18n/en.i18n.json]: Translation maintained.

File: packages/rocketchat-i18n/i18n/en.i18n.json (hunk 4 above)
Comment (vickyokrm) [reply packages/rocketchat-i18n/i18n/en.i18n.json]: translation maintained.

File: packages/rocketchat-i18n/i18n/en.i18n.json (hunk 5 above)
Comment (vickyokrm) [reply packages/rocketchat-i18n/i18n/en.i18n.json]: Translation Maintained.

File: packages/rocketchat-message-attachments/client/fieldTypeJoinRoomRequest.js (hunk 1 above)
Comment (vickyokrm) [reply packages/rocketchat-message-attachments/client/fieldTypeJoinRoomRequest.js]: adapted.

File: packages/rocketchat-ui/client/views/app/directory.js (hunk 3 above)
Comment (vickyokrm) [reply packages/rocketchat-ui/client/views/app/directory.js]: adapted the `view-outside-room` permission

File: packages/rocketchat-lib/server/models/Messages.js [hunk 9]
```diff
@@ -318,10 +320,34 @@ RocketChat.models.Messages = new class extends RocketChat.models._Base {
 		return this.insert(record);
 	}
 
+	findJoinRequestsByMsgTypeAndRoom(msgType, rid) {
```
Comment (mrsimpson) [inline packages/rocketchat-lib/server/models/Messages.js]: @vickyokrm still not nice: if there's a `msgType` in the  signature, `JoinRequest` it must not be part of the name. Since it's a method of the message-model, it should be `findByTypeAndRoom`

File: packages/rocketchat-lib/server/models/Messages.js [hunk 10]
```diff
@@ -318,10 +320,34 @@ RocketChat.models.Messages = new class extends RocketChat.models._Base {
...
+		const query = {t: msgType, rid};
+		const options = {
+			sort: {
+				ts: -1
+			}
+		};
+		return this.find(query, options);
+	}
+
+	findLatestJoinRequestByMsgTypeRoomAndUser(msgType, rid, userName) {
```
Comment (mrsimpson) [inline packages/rocketchat-lib/server/models/Messages.js]: see above:  `findLatestByTypeAndRoom`

**Comment by mrsimpson** [comment]
- [ ] Topics andRequests should also be available in 
- [ ] The icons of the rom types in directory are not used (hash vs. lock)
- [ ] Explanatory text for users that are not allowed to add the user
- [ ] There should be a setting which room types are displayed in directory

Further enhancement: Explicitly hide some rooms from the directory

**Comment by ThomasRoehl** [comment]
Tests have been added to validate "accept" and "decline" options and finding the private room via the new list

File: packages/rocketchat-message-attachments/client/stylesheets/messageAttachments.css [hunk 11]
```diff
@@ -144,11 +144,18 @@ html.rtl .attachment {
 	  	font-weight: 500;
 	}
 
+	& .join-room-request {
+	    color: black;
+    }
+
+    & .join-room-request-reason {
```
Comment (mrsimpson) [inline packages/rocketchat-message-attachments/client/stylesheets/messageAttachments.css]: check indention

**Comment by janrudolph** [comment]
Just a short question @vickyokrm : Are you able to request access to a archived channel? This should not be possible, or what do you think @mrsimpson ?

**Comment by mrsimpson** [comment]
Well, why not? You‘d be able to read what happened. But is that actually possible?

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

File: server/methods/browseChannels.js [hunk 12]
```diff
@@ -24,18 +24,20 @@ const sortUsers = function(field, direction) {
...
 Meteor.methods({
-	browseChannels({text='', type = 'channels', sortBy = 'name', sortDirection = 'asc', page = 0, limit = 10}) {
+	browseChannels({text = '', type = 'Channels', sortBy = 'name', sortDirection = 'asc', page = 0, limit = 10}) {
 		const regex = new RegExp(s.trim(s.escapeRegExp(text)), 'i');
-
-		if (!['channels', 'users', 'private'].includes(type)) {
-			return;
+		let roomType;
+		if (type !== 'users') {
+			roomType = Object.keys(RocketChat.roomTypes.roomTypes).find(key => RocketChat.roomTypes.roomTypes[key]._label === type);
```
Comment (vickyokrm) [inline server/methods/browseChannels.js]: Passing a value for a new `roomType` parameter will be another way to avoid this computation of room type again.

**Comment by ThomasRoehl** [comment]
**Tests:**
- entering private channel, requests and topics via the 'all rooms' tab
- ask invite for private channel
- decline and accept invitation
- automated tests are running through

**Change requests:**
- if you are not a member of a request or topic shown in the list and you want to enter the room, you end up on the "room not found" page. There should be a different message or the same "ask for invitation" as with the private channels.
- If someone asks for an invitation to a private channel, the message gets send to the room, as an owner/admin etc. you don't get a new message notification

File: packages/rocketchat-lib/server/models/Messages.js (hunk 9 above)
Comment (mrsimpson) [reply packages/rocketchat-lib/server/models/Messages.js]: @vickyokrm this is still pending, right?

File: packages/assistify-help-request/server/migrations.js [hunk 13]
```diff
@@ -13,19 +13,38 @@ const removeObsoleteSubscriptions = function() {
...
 
+function makePrivateRoomsSecret() {
+	let count = 0;
+	RocketChat.models.Rooms.findByType('p').forEach((room) => {
+		if ('customFields' in room) {
+			if ('secretRoom' in room.customFields) {
+				return;
+			}
+		}
+		RocketChat.models.Rooms.setPrivacyById(room._id, true);
```
Comment (mrsimpson) [inline packages/assistify-help-request/server/migrations.js]: In the migration, you don't necessarily need to use a method. You can set it in one DB statement: `update({t: 'p', hidden: {$exists: false}}, {$set: {hidden: true}}, {multi: true})`

File: packages/assistify-help-request/server/migrations.js [hunk 14]
```diff
@@ -13,19 +13,38 @@ const removeObsoleteSubscriptions = function() {
...
+		RocketChat.models.Subscriptions.model.remove({_id: errSubscription._id.toString()});
 		counter++;
 	});
 	return counter;
 };
 
+function makePrivateRoomsSecret() {
+	let count = 0;
+	RocketChat.models.Rooms.findByType('p').forEach((room) => {
+		if ('customFields' in room) {
```
Comment (mrsimpson) [inline packages/assistify-help-request/server/migrations.js]: I'd favor not to use custom fields but to enhance the model (e. g. with a `hidden` property)

File: packages/rocketchat-channel-settings/client/views/channelSettings.html [hunk 15]
```diff
@@ -140,103 +149,129 @@
 				</div>
 			</div>
 		</div>
-		{{/if}} {{/with}} {{#with settings.archived}} {{#if canView}}
+	{{/if}} {{/with}} {{#with settings.privacy}} {{#if settings.t.showPrivacySetting}}
```
Comment (mrsimpson) [inline packages/rocketchat-channel-settings/client/views/channelSettings.html]: I think the term "private" is already taken.

File: packages/rocketchat-channel-settings/client/views/channelSettings.js [hunk 16]
```diff
@@ -54,7 +54,7 @@ Template.channelSettingsEditing.events({
 });
 
 Template.channelSettingsEditing.onCreated(function() {
-	const room = this.room = ChatRoom.findOne(this.data && this.data.rid);
+	const room = this.room = ChatRoom.findOne({_id: this.data.rid});
```
Comment (mrsimpson) [inline packages/rocketchat-channel-settings/client/views/channelSettings.js]: why this change?

File: packages/rocketchat-channel-settings/client/views/channelSettings.js [hunk 17]
```diff
@@ -177,14 +177,20 @@ Template.channelSettingsEditing.onCreated(function() {
...
 				return (RocketChat.authz.hasAllPermission('edit-room', room._id) && !room['default']) || RocketChat.authz.hasRole(Meteor.userId(), 'admin');
 			},
 			save(value) {
 				const saveRoomSettings = () => {
 					value = value ? 'p' : 'c';
 					RocketChat.callbacks.run('roomTypeChanged', room);
-					return call('saveRoomSettings', room._id, 'roomType', value).then(() => {
+					return call('saveRoomSettings', room._id, {
+						'roomType': value,
+						'secretRoom': value === 'p' ? room.customFields.secretRoom : false
```
Comment (mrsimpson) [inline packages/rocketchat-channel-settings/client/views/channelSettings.js]: The defaulting should not happen on the UI. instead, the method creating the room should handle it.

File: packages/rocketchat-channel-settings/server/functions/saveRoomPrivacy.js [hunk 18]
```diff
@@ -0,0 +1,21 @@
...
+		});
+	}
+	const room = RocketChat.models.Rooms.findOneById(rid);
+	if (room == null) {
+		throw new Meteor.Error('error-invalid-room', 'error-invalid-room', {
+			'function': 'RocketChat.saveRoomType',
+			_id: rid
+		});
+	}
+	if (room.t === 'd') {
```
Comment (mrsimpson) [inline packages/rocketchat-channel-settings/server/functions/saveRoomPrivacy.js]: copy-paste-artifact?

File: packages/rocketchat-i18n/i18n/de.i18n.json [hunk 19]
```diff
@@ -1662,6 +1664,10 @@
   "Report_sent": "Bericht gesendet",
   "Report_this_message_question_mark": "Diese Nachricht melden?",
   "Reporting": "Berichtswesen",
+  "Request": "Anfrage",
```
Comment (mrsimpson) [inline packages/rocketchat-i18n/i18n/de.i18n.json]: This should be part of the assistify-file

File: packages/rocketchat-lib/server/methods/joinRoom.js [hunk 20]
```diff
@@ -18,18 +19,79 @@ Meteor.methods({
...
+			throw new Meteor.Error('error-invalid-user', 'Invalid user', {method: 'requestJoin'});
+		}
+
+		const room = RocketChat.models.Rooms.findOneByName(name);
+		if (!room) {
+			throw new Meteor.Error('error-invalid-room', 'Invalid room', {method: 'requestJoin'});
+		}
+
+		const rocketCatUser = RocketChat.models.Users.findOneByUsername('rocket.cat');
+		if (rocketCatUser && user) {
```
Comment (mrsimpson) [inline packages/rocketchat-lib/server/methods/joinRoom.js]: What about an `else`?

File: packages/rocketchat-lib/server/models/Rooms.js [hunk 21]
```diff
@@ -667,6 +683,16 @@ class ModelRooms extends RocketChat.models._Base {
 		return this.update(query, update);
 	}
 
+	setPrivacyById(_id, value) {
```
Comment (mrsimpson) [inline packages/rocketchat-lib/server/models/Rooms.js]: `setSecrecyById`?

**Comment by mrsimpson** [comment]
@vickyokrm I‘m not fully done yet. Just wanted to comment what I’ve seen so far

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

File: packages/rocketchat-ui/client/views/app/directory.html [hunk 22]
```diff
@@ -1,87 +1,92 @@
...
-						<input type="text" class="rc-input__element rc-input__element--small js-search" name="message-search" id="message-search" placeholder="{{_ 'Directory_Search_Placeholder'}}" autocomplete="off">
-					</div>
-				</label>
+			<div class="rc-directory-header">
+				<select class="rc-input__element rc-directory-selector js-typeSelector" name="type">
+					<option value="Channels">{{_ 'Channels'}}</option>
+					<option value="users">{{_ 'Users' }}</option>
+					<option value="Private_Groups">{{_ 'Private' }}</option>
+					{{#each roomType in customRoomTypes}}
+						<option value={{roomType}}> {{roomType}}</option>
```
Comment (mrsimpson) [inline packages/rocketchat-ui/client/views/app/directory.html]: `{{roomType}}` should be an i18n-supported text. I didn't see this in the helper. At my client, the dropdown value was not translated

File: packages/rocketchat-lib/server/models/Rooms.js [hunk 23]
```diff
@@ -256,6 +256,19 @@ class ModelRooms extends RocketChat.models._Base {
 		return this._db.find(query, options);
 	}
 
+	findByNameAndTypeAndNotSecret(name, type, options) {
```
Comment (mrsimpson) [inline packages/rocketchat-lib/server/models/Rooms.js]: I noticed that this semantic is not sufficient: Even if a room is secret, it should be listed if the user is a member of this room. Thus, the semantic better should be `findListableByNameAndType`

File: packages/rocketchat-ui/client/views/app/directory.js [hunk 24]
```diff
@@ -31,6 +22,13 @@ function directorySearch(config, cb) {
 					createdAt: timeAgo(result.createdAt)
 				};
 			}
+			return {
```
Comment (mrsimpson) [inline packages/rocketchat-ui/client/views/app/directory.js]: wrapping this in an `else` would be more elegant

File: packages/rocketchat-ui/client/views/app/directory.html [hunk 25]
```diff
@@ -1,87 +1,92 @@
...
+					</label>
+				</div>
+				<button class="rc-button rc-button--small rc-button--primary rc-directory-plus">{{> icon icon="plus" }}</button>
 			</div>
-			<button class="rc-button rc-button--small rc-button--primary rc-directory-plus">{{> icon icon="plus" }}</button>
-		</div>
 		{{/header}}
 		<div class="rc-directory-content">
-			{{#if $eq searchType 'channels'}}
+			{{#if $eq searchType 'users'}}
```
Comment (mrsimpson) [inline packages/rocketchat-ui/client/views/app/directory.html]: From a usability perspective, I'd wish for a note on top of the list, that - if rooms are secret - secret rooms will not be listed here.
Else, particularly after the first migration after which all rooms are secret, user's would glance at an empty list and not understand it.
Proposal:

```
{{#if secretRoomsExist}}
<div class=something-like-announcement>{{_ secret_rooms_not_listed}}<div>
{{/if}}
```

File: packages/assistify-help-request/server/migrations.js (hunk 13 above)
Comment (vickyokrm) [reply packages/assistify-help-request/server/migrations.js]: Handled using the Base model `update` method.

File: packages/assistify-help-request/server/migrations.js (hunk 14 above)
Comment (vickyokrm) [reply packages/assistify-help-request/server/migrations.js]: Extended the model schema with a new field `secret` instead of using the custom Fields.

File: packages/rocketchat-channel-settings/client/views/channelSettings.js (hunk 16 above)
Comment (vickyokrm) [reply packages/rocketchat-channel-settings/client/views/channelSettings.js]: This change is reverted as this change does not make any difference.

File: packages/rocketchat-i18n/i18n/de.i18n.json [hunk 26]
```diff
@@ -1065,6 +1065,7 @@
   "join-without-join-code_description": "Berechtigung, Kanäle mit Zutrittscode auch ohne Code zu betreten",
   "join-without-join-code": "Ohne Code beitreten",
   "join": "Beitreten",
+  "Join-room-request": "__user__ möchte diesem Raum als Mitglied beitreten.  Besitzer oder Moderator sollten darauf antworten",
```
Comment (vickyokrm) [inline packages/rocketchat-i18n/i18n/de.i18n.json]: All these translations are relevant to this feature and should be contributed back to RC. so no change required.

File: packages/rocketchat-i18n/i18n/de.i18n.json (hunk 19 above)
Comment (vickyokrm) [reply packages/rocketchat-i18n/i18n/de.i18n.json]: These changes must be contributed to RC.

File: packages/rocketchat-lib/server/methods/joinRoom.js (hunk 20 above)
Comment (vickyokrm) [reply packages/rocketchat-lib/server/methods/joinRoom.js]: exception will be raised upon failure.

File: packages/rocketchat-lib/server/models/Rooms.js (hunk 23 above)
Comment (vickyokrm) [reply packages/rocketchat-lib/server/models/Rooms.js]: With this change, the private channel -> marked as `secret` will be displayed on the search, if the user is part of the channel and not otherwise.

File: packages/rocketchat-message-attachments/client/stylesheets/messageAttachments.css (hunk 11 above)
Comment (vickyokrm) [reply packages/rocketchat-message-attachments/client/stylesheets/messageAttachments.css]: Fixed.

File: packages/rocketchat-ui/client/views/app/directory.html (hunk 22 above)
Comment (vickyokrm) [reply packages/rocketchat-ui/client/views/app/directory.html]: hanlded.

File: packages/rocketchat-ui/client/views/app/directory.html (hunk 25 above)
Comment (vickyokrm) [reply packages/rocketchat-ui/client/views/app/directory.html]: WIP ;)

File: packages/rocketchat-ui/client/views/app/directory.js (hunk 24 above)
Comment (vickyokrm) [reply packages/rocketchat-ui/client/views/app/directory.js]: handled with else block.

**Comment by mrsimpson** [comment]
Doesn't work. 

- Created admin + two users
- Create private room with first one - "secret" not selected
- The room is listed for admin + first, but not for second one.

I checked on the database, `secret` is `false`

**Comment by vickyokrm** [comment]
@mrsimpson Machi, it works for me. Lets check together on your side.

**Comment by mrsimpson** [comment]
This was due to the second user being a `guest`. That's fine.

other findings 

- [ ] In the room settings, the description about what a secret room is should be visible as well
- [ ] Banner is shown always - though all rooms are listed
- [ ] Notification for the admin that the second user wants to join is not created

**Comment by vickyokrm** [comment]
- [ ] Channel name highlight for the new join request system message will be  checked based on user permission
- [ ] Channel settings like Private<->Public and Read only is disabled for assistify channels
- [ ] Minor bugs and translation issues.
- [ ] Hinweis for the secret channel is added under the channel settings ( The german text for the hint is wrapped)

**Comment by janrudolph** [comment]
- German translation is not understandable "Der Kanal wird nur Mitgliedern im Verzeichnis aller Räume angezeigt" => Es ist nicht klar, auf das sich die Mitglieder beziehen. Besser wäre "Der Kanal wird nur Mitgliedern diesen Raumes im Verzeichnis aller Räume angezeigt"

**Comment by ruKurz** [comment]
### Request membership for a channel

#### Access denied view
image

**General**

- The site is very daunting.
- The text is too general. It is about a channel / room / group.

**Color improvements**

- color: blue (info) or orange (warning)

**German text improvements**
- Der Zugriff auf den Kanal „XYZ“ ist geschützt.
- Du darfst den Kanal „XYZ“ nicht sehen.

**English text improvements**

- Access to the "XYZ" channel is protected.
- You may not see the channel "XYZ".


#### Open/Start a request

image

**German text improvements**

- Text: Möchtest Du dem Kanal „XYZ“ beitreten? Frage einfach den Besitzer/Moderator, ob Du beitreten darfst.
- Button: „Beitrittsanfrage senden“

**English text improvements**

- Text: Would you like to join the "XYZ" channel? Just ask the owner/moderator if you can join.
- Button: Send membership application


#### Request Dialog

image

**German text improvements**

- Titel: Du möchtest dem Kanal „XYZ“ beitreten? 
- Beschreibung: Am besten gibst Du einen Grund an, sodass der Eigentümer/Moderator versteht, warum Du beitreten möchtest.
- Button: Anfrage zum Beitritt senden.

**English text improvements**

- Title: You want to join the channel "XYZ"? 
- Description: It is best to give a reason so that the owner/moderator understands why you want to join.
- button: Send a request to join.


### Edit Inquiry

After the user has made a request, a message appears in the "requested" channel. However, the affected channel is not marked as unread for the owner/moderator, so the processing of the request may remain unreadable because the person responsible does not see the request. If necessary, it makes sense to inform all owners/moderators about the request by e-mail to ensure that the request is processed.
￼
**Permission dependent information**

Due to the permissions, this message should only be displayed for those users who are allowed to accept or reject the request. This eliminates the text "Owners or moderators should reply to it."

**German**

Korrekturtext: Besitzer oder Moderatoren sollten darauf antworten.
(Moderatoren —> Plural und Punkt am Ende)

**English**

Proofreading: Owners or moderators should respond.
(Moderators -> plural and dot at the end)


### Reject/accept inquiry

If a request has been rejected/accepted, I will not receive any information about it.

**tweaks**
- If the request has been processed, the applicant should be informed in both cases. Additionally also by mail.
- Just as the applicant should state a reason for admission, the application processor should also state a reason for rejection (eye level and so on).

**Comment by ruKurz** [comment]
I've now separated all my feedback into several issues:

- https://github.com/RocketChat/Rocket.Chat/issues/11294
- https://github.com/assistify/Rocket.Chat/issues/377
- https://github.com/assistify/Rocket.Chat/issues/378
- https://github.com/assistify/Rocket.Chat/issues/379
- https://github.com/assistify/Rocket.Chat/issues/380
